### PR TITLE
[codenav]: refactor of auto-indexing background jobs

### PR DESCRIPTION
**Pull request by Numbers88s**
Refactor of auto-indexing background jobs as part of our ongoing efforts around 619

## Test plan
Make sure all test pass

woooo

### CODE REVIEW

File: internal/codeintel/uploads/internal/store/store_repositories.go [hunk 1]
```diff
@@ -14,6 +15,103 @@ import (
...
+		quote(table),
+		now,
+		now,
+	)))
+}
+
+func quote(s string) *sqlf.Query { return sqlf.Sprintf(s) }
+
+const selectRepositoriesForIndexScanQuery = `
+-- source: internal/codeintel/stores/dbstore/uploads.go:selectRepositoriesForIndexScan
```
Comment (efritz) [inline internal/codeintel/uploads/internal/store/store_repositories.go:55]: Note for later cleanup: we'll want to ensure all these sources are correct (unless we can automate their inclusion across the entire repo, cc @jhchabran)

**Review by efritz — APPROVED** [review]

File: internal/codeintel/uploads/service.go [hunk 2]
```diff
@@ -308,6 +310,20 @@ func (s *Service) SetRepositoriesForRetentionScan(ctx context.Context, processDe
...
 
+func (s *Service) GetRepoName(ctx context.Context, repositoryID int) (_ string, err error) {
+	ctx, _, endObservation := s.operations.getRepoName.With(ctx, &err, observation.Args{})
+	defer endObservation(1, observation.Args{})
+
+	return s.store.RepoName(ctx, repositoryID)
+}
+
+func (s *Service) GetRepositoriesForIndexScan(ctx context.Context, table, column string, processDelay time.Duration, allowGlobalPolicies bool, repositoryMatchLimit *int, limit int, now time.Time) (_ []int, err error) {
+	ctx, _, endObservation := s.operations.getRepositoriesForIndexScan.With(ctx, &err, observation.Args{})
```
Comment (Strum355) [inline internal/codeintel/uploads/service.go:321]: Important to add some/all of the parameters to `observation.Args` here so we can see them in traces etc

File: internal/codeintel/autoindexing/transport/graphql/resolver.go [hunk 3]
```diff
@@ -2,86 +2,138 @@ package graphql
...
+		return nil, err
+	}
+
+	return hints, nil
 }
 
-func (r *Resolver) UpdateRepositoryIndexConfiguration(ctx context.Context, args *gql.UpdateRepositoryIndexConfigurationArgs) (_ *gql.EmptyResponse, err error) {
-	ctx, _, endObservation := r.operations.updateRepositoryIndexConfiguration.With(ctx, &err, observation.Args{})
+func (r *resolver) UpdateIndexConfigurationByRepositoryID(ctx context.Context, repositoryID int, configuration string) (err error) {
+	ctx, _, endObservation := r.operations.updateIndexConfigurationByRepositoryID.With(ctx, &err, observation.Args{})
```
Comment (Strum355) [inline internal/codeintel/autoindexing/transport/graphql/resolver.go:127]: same here, pls add params to `observation.Args`

File: internal/codeintel/autoindexing/internal/store/scan.go [hunk 4]
```diff
@@ -1,15 +1,187 @@
...
+	defer func() { err = basestore.CloseRows(rows, err) }()
+
+	for rows.Next() {
+		if err := rows.Scan(&value); err != nil {
+			return 0, err
+		}
+	}
+
+	return value, nil
+}
```
Comment (Strum355) [inline internal/codeintel/autoindexing/internal/store/scan.go:187]: As a drive-by note, we can probably remove use of this and replace it with `basestore.ScanInt` instead, some examples [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c82f59bcab1b5a5c93ad12202da252188636ac59/-/blob/internal/database/basestore/rows.go?L213:2#tab=references), and [heres](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c82f59bcab1b5a5c93ad12202da252188636ac59/-/blob/internal/codeintel/autoindexing/internal/store/scan.go?L174:6#tab=references) the two places where `scanCount` is used

File: internal/codeintel/autoindexing/internal/store/scan.go [hunk 5]
```diff
@@ -1,15 +1,187 @@
...
+		var count int
+		if err := rows.Scan(&id, &count); err != nil {
+			return nil, err
+		}
+
+		visibilities[id] = count
+	}
+
+	return visibilities, nil
+}
```
Comment (Strum355) [inline internal/codeintel/autoindexing/internal/store/scan.go:135]: Replacable with [basestore.NewMapScanner](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c82f59bcab1b5a5c93ad12202da252188636ac59/-/blob/internal/database/basestore/rows.go?L117:6-117:19)

File: internal/codeintel/autoindexing/background/scheduler/scheduler.go [hunk 6]
```diff
@@ -2,26 +2,125 @@ package scheduler
...
+				err = errors.Append(err, repositoryErr)
+			}
+		}
+	}
+
+	return err
+}
+
+func (s *scheduler) handleRepository(ctx context.Context, repositoryID int, now time.Time) error {
+	offset := 0
```
Comment (Strum355) [inline internal/codeintel/autoindexing/background/scheduler/scheduler.go:82]: Im more a fan of `var offset int` for cases like this, but thats completely pedantic :p

File: internal/codeintel/autoindexing/internal/store/store_indexes.go [hunk 7]
```diff
@@ -5,11 +5,362 @@ import (
...
+	docker_steps,
+	local_steps,
+	root,
+	indexer,
+	indexer_args,
+	outfile,
+	execution_logs
+) VALUES %s
+RETURNING id
+`
```
Comment (Strum355) [inline internal/codeintel/autoindexing/internal/store/store_indexes.go:86]: I think you need to rebase main onto your branch, it looks like this already exists on main, or else im getting confused with something here :thinking:

**Review by Strum355 — COMMENTED** [review]
Just some polish around `observation.Args` and db scanners stuff :+1:

File: internal/codeintel/autoindexing/background/scheduler/scheduler.go (hunk 6 above)
Comment (Numbers88s) [reply internal/codeintel/autoindexing/background/scheduler/scheduler.go:82]: Cool good to know! This is not my code leaving it as is for now, the scope of this PR is to move stuff to the correct layer. It would complicate the diff if I also update code. I can circle back and address those later to make the code review a little easier. Good to know tho man. Use default values instead of making it explicit :thumbsup:

File: internal/codeintel/autoindexing/internal/store/scan.go (hunk 5 above)
Comment (Numbers88s) [reply internal/codeintel/autoindexing/internal/store/scan.go:135]: Cool cool, leaving as if for now, the next iteration is to collapse some of the code so I'll address it then. Also just briefly looking at it that `basestore.NewMapScanner` is abstact as hell! A bit hard to follow for me.

File: internal/codeintel/autoindexing/internal/store/scan.go (hunk 4 above)
Comment (Numbers88s) [reply internal/codeintel/autoindexing/internal/store/scan.go:187]: yup, same as before leaving as is for now

File: internal/codeintel/autoindexing/transport/graphql/resolver.go (hunk 3 above)
Comment (Numbers88s) [reply internal/codeintel/autoindexing/transport/graphql/resolver.go]: Done, added it.

File: internal/codeintel/uploads/service.go (hunk 2 above)
Comment (Numbers88s) [reply internal/codeintel/uploads/service.go]: Done, added it!

File: internal/codeintel/uploads/internal/store/store_repositories.go (hunk 1 above)
Comment (Numbers88s) [reply internal/codeintel/uploads/internal/store/store_repositories.go]: Fixed the ones on this file. I'll do a wider pass and update if I need to 👍
